include/Mat.h

```diff
@@ -453,100 +459,118 @@ class Mat {
 
     template<class Type2, class Type3>
     Mat<Type3> broadcast(const Mat<Type2> &b, Type3 (*f)(Type, Type2)){
-        if(ndims == 2){
-            int r,c;
-            if(dims[0] >= b.dims[0]) r = dims[0];
-            else r = b.dims[0];
-            if(dims[1] >= b.dims[1]) c = dims[1];
-            else c = b.dims[1];
-            Mat<Type3> out(r,c);
-            broadcast(b, f, out);
-            return out;
-        }
-        else if(ndims == 1){
-            int c;
-            if(dims[0] >= b.dims[0]) c = dims[0];
-            else c = b.dims[0];
-            Mat<Type3> out(c);
-            broadcast(b, f, out);
-            return out;
-        }
-        else errorCheck(true, "n-dimensional broadcast not yet implemented");
-        return Mat<Type3>::zeros(0);
+        Mat<Type3> out;
+        delete[] out.dims;
+
+        size_type* big_dim;
+        size_type* small_dim;
+        long dimdiff;
+        if(ndims >= b.ndims){
+            big_dim = dims;
+            small_dim = b.dims;
+            dimdiff = ndims - b.ndims;
+            out.ndims = ndims;
+            out.dims = new size_type[ndims];
+            out.strides = new size_type[ndims];
+        }
+        else{
+            big_dim = b.dims;
+            small_dim = dims;
+            dimdiff = b.ndims - ndims;
+            out.ndims = b.ndims;
+            out.dims = new size_type[b.ndims];
+            out.strides = new size_type[b.ndims];
+        }
+
+        for(long i = 0; i < out.ndims; i++){
+            if(i < dimdiff) out.dims[i] = big_dim[i];
+            else{
+                errorCheck(big_dim[i] != small_dim[i - dimdiff] &&
+                            big_dim[i] != 1 && small_dim[i - dimdiff] != 1,
+                            "operand frames not aligned");
+                if(big_dim[i] >= small_dim[i - dimdiff]) out.dims[i] = big_dim[i];
+                else out.dims[i] = small_dim[i - dimdiff];
+            }
+        }
+        out.buildStrides();
+
+        delete[] out.memory;
+        out.memory = new Type3[out.size()];
+        out.data = out.memory;
+
+        broadcast(b, f, out);
+        return out;
     }
 
     template<class Type2, class Type3>
     void broadcast(const Mat<Type2> &b, Type3 (*f)(Type, Type2), Mat<Type3> &out){
-        for(long n = 0; n < ndims; n++){
-            if(dims[n] > b.dims[n]){
-                errorCheck(b.dims[n] != 1 || out.dims[n] != dims[n],
+        if(ndims >= b.ndims){
```

Comment:
check that the out matrix is the right shape for holding the bigger of the aligned dimensions